pygbm/loss.py

```diff
@@ -104,5 +114,82 @@ def _update_gradients_hessians_logistic(gradients, hessians, y_true, y_pred):
+        # reshape y_pred to (n_samples, n_trees_per_iteration)
+        y_pred = y_pred.reshape(self.n_trees_per_iteration, -1).T
+        return _update_gradients_hessians_softmax(gradients, hessians,
+                                                  y_true, y_pred,
+                                                  self.n_trees_per_iteration)
+
+    def predict_proba(self, decision):
+        # TODO: This could be done in parallel
+        decision = decision.reshape(self.n_trees_per_iteration, -1).T
+        # compute softmax (using exp(log(softmax)))
+        proba = np.exp(decision - logsumexp(decision, axis=1)[:, np.newaxis])
+        return proba
+
+
+@njit(parallel=True)
+def _update_gradients_hessians_softmax(gradients, hessians, y_true, y_pred,
+                                       n_trees_per_iteration):
+    # gradients, hessians, and y_pred are of shape
+    # (n_samples * n_trees_per_iteration,).
+    # y_true is of shape (n_samples,).
+    #
+    # instead of passing the whole gradients and hessians arrays and slicing
+    # them here, we could instead do the update in the 'for k in ...' loop of
+    # fit(), by passing gradients_at_k and hessians_at_k which are of size
+    # (n_samples,).
+    # That would however require to pass a copy of y_pred, so it does not get
+    # partially overwritten at the end of the loop when _update_y_pred() is
+    # called.
+    n_samples = y_true.shape[0]
+    starts, ends, n_threads = _get_threads_chunks(total_size=n_samples)
+    for k in range(n_trees_per_iteration):
+        gradients_at_k = gradients[n_samples * k:n_samples * (k + 1)]
+        hessians_at_k = hessians[n_samples * k:n_samples * (k + 1)]
+        for thread_idx in prange(n_threads):
+            for i in range(starts[thread_idx], ends[thread_idx]):
+                # p_k is the probability that class(ith sample) == k.
+                # This is a regular softmax. Ideally we should be using
+                # logaddexp or logsumexp but they are not supported by numba.
+                denominator = np.sum(np.exp(y_pred[i, :]))
+                p_k = np.exp(y_pred[i, k] - np.log(denominator))
+                gradients_at_k[i] = p_k - (y_true[i] == k)
+                hessians_at_k[i] = p_k * (1. - p_k)
+                # LightGBM uses 2 * p_k * (1 - p_k) but that seems incorrect?
+                # Looks like it gives much better results in practice though.
+                # Both versions pass the compare_lgbm tests but that's because
+                # we only build 1 tree.
+                # lightgbm version does not pass the numerical hessian test.
+
+                # Note: I observed super weird numba bugs here.
+                # This fails:
+                # p_k = np.exp(y_pred[i, k] - np.log(np.sum(np.exp(y_pred[i]))
+                # This fails as well
+                # p_k = np.exp(y_pred[i, k] -
+                #              np.log(np.sum(np.exp(y_pred[i, :]))
```

Comment:
Shouldn't we use a numerically stable primitive such as the code in `scipy.special.logsumexp`?

https://github.com/scipy/scipy/blob/master/scipy/special/_logsumexp.py#L9-L129

This is not a primitive in numpy but the code is pure numpy so numba should be able to optimize it in nopython mode although I have not checked. If that's not the case we will have to maintain our own numba-compatible implementation of logsumexp in pygbm.

Reply:
I made a comment about this a few lines above in the code.

The function we need is `scipy.logsumexp` but it's not supported. `np.logaddexp.reduce` is equivalent, but numba does not support `reduce` it seems.


```python
from numba import njit
import numpy as np

@njit
def g():
    a = np.arange(10).astype(np.float32)
    b = np.logaddexp.reduce(a)
    return b

print(g())

```

```
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'reduce' of type Function(<ufunc 'logaddexp'>)

```

Reply:
But the code of logsumexp is only doing basic numpy operation so it should work in numba nopython mode, no? If not we could / should adapt our own copy of it in pygbm to make it run in `@njit` mode

Reply:
I'm not sure that numba supports functions defined in scipy. We had the same issue with `expit`.


```
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'logsumexp': cannot determine Numba type of <class 'function'>

File "numba_lol.py", line 8:
def g():
    <source elided>
    a = np.arange(10).astype(np.float32)
    b = logsumexp(a)
    ^
```

You mean we should have our own `logaddexp` **with** numerical stability?
Because we already have the basic (potentially unstable) version here in the code

Reply:
You need to wrap it with the `njit` decorator before calling it from an `njit`ed function.

Reply:
Like this you mean? Still getting the error (all functions are njitted even though it doesn't show on the trace)

```
Untyped global name 'logsumexp': cannot determine Numba type of <class 'function'>

File "numba_lol.py", line 7:
def _logsumexp(x):
    return logsumexp(x)
    ^

[1] During: resolving callee type: type(CPUDispatcher(<function _logsumexp at 0x7f761c3951e0>))
[2] During: typing of call at numba_lol.py (12)


File "numba_lol.py", line 12:
def g():
    <source elided>
    a = np.arange(10).astype(np.float32)
    b = _logsumexp(a)
```

The numerical stability trick for logsumexp is [a classic](http://cs231n.github.io/linear-classify/#softmax) I think, I just pushed something.